### PR TITLE
Implement max-length config for TextArea input type #5336

### PR DESCRIPTION
**Pull request by vbradnitski**

### CODE REVIEW

**Comment by vbradnitski** [comment]
Contains changes from #5335 
@alansemenov TextLine contains 'data-regex-status', which displays 'Valid/Invalid' text depending on regexp check. But with new validation it looks a bit not very obvious in my opinion. Should it be changed?
image

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/text/TextLine.ts [hunk 1]
```diff
@@ -6,21 +6,28 @@ module api.form.inputtype.text {
...
 
         private readConfig(inputConfig: { [element: string]: { [name: string]: string }[]; }): void {
-            let regexpConfig = inputConfig && inputConfig['regexp'] && inputConfig['regexp'][0];
-            let regexp = regexpConfig && regexpConfig['value'];
-            this.regexpStr = regexp || null;
+            const regexpConfig = inputConfig['regexp'] ? inputConfig['regexp'][0] : {};
+            const regexp = regexpConfig ? regexpConfig['value'] : '';
+            this.regexp = new RegExp(regexp);
+
+            const maxLengthConfig = inputConfig['max-length'] ? inputConfig['max-length'][0] : {};
```
Comment (alansemenov) [inline modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/text/TextLine.ts]: ``max-length`` property from config xml gets converted to ``maxLength`` on the server, so this should be changed to ``inputConfig['maxLength']``

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/text/TextInputType.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        private maxLength: number;
+
+        constructor(config: api.form.inputtype.InputTypeViewContext) {
+            super(config);
+            this.readConfig(config.inputConfig);
+        }
+
+        protected readConfig(inputConfig: { [element: string]: { [name: string]: string }[]; }): void {
+            const maxLengthConfig = inputConfig['max-length'] ? inputConfig['max-length'][0] : {};
```
Comment (alansemenov) [inline modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/text/TextInputType.ts]: ``max-length`` property from config xml gets converted to ``maxLength`` on the server, so this should be changed to ``inputConfig['maxLength']``

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/text/TextInputType.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                    inputEl.getParentElement().appendChild(lengthCounter);
+                }
+            });
+
+            return inputEl;
+        }
+
+        private updateLengthCounterValue(lengthCounter: DivEl, newValue: string) {
+            lengthCounter.setHtml(`${this.maxLength - newValue.length}`);
+        }
```
Comment (alansemenov) [inline modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/text/TextInputType.ts]: The counter should not go below 0 according to the task description, but now it does.

File: modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/text/TextInputType.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        protected readConfig(inputConfig: { [element: string]: { [name: string]: string }[]; }): void {
+            const maxLengthConfig = inputConfig['maxLength'] ? inputConfig['maxLength'][0] : {};
+            const maxLength = NumberHelper.toNumber(maxLengthConfig['value']);
+            this.maxLength = maxLength > 0 ? maxLength : null;
+        }
+
+        protected initOccurenceListeners(inputEl: FormInputEl) {
+
+            inputEl.onValueChanged((event: api.ValueChangedEvent) => {
+                if (NumberHelper.isNumber(this.maxLength)) {
```
Comment (alansemenov) [inline modules/admin/admin-ui/src/main/resources/web/admin/common/js/form/inputtype/text/TextInputType.ts]: Value in ``this.maxLength`` will be set only once and never change, so there's no point in checking for ``isNumber`` every time inside these two handlers. Instead, we should create both event handlers only **if** ``this.maxLength`` is a number:

```Typescript
if (NumberHelper.isNumber(this.maxLength)) {
                inputEl.onValueChanged((event: api.ValueChangedEvent) => {
                    const lengthCounter = Element.fromHtmlElement(
                        (<HTMLElement>inputEl.getParentElement().getHTMLElement().querySelector('.length-counter')));
                    if (lengthCounter) {
                        this.updateLengthCounterValue(lengthCounter, inputEl.getValue());
                    }
                });

                inputEl.onRendered(() => {
                    const lengthCounter = new DivEl('length-counter');
                    this.updateLengthCounterValue(lengthCounter, inputEl.getValue());

                    inputEl.getParentElement().appendChild(lengthCounter);
                });
            }
```

**Pull request merged**